src/treetensornetworks/opsum_to_ttn/opsum_to_ttn.jl

```diff
@@ -88,130 +77,119 @@ function pos_in_link!(linkmap::Dict, k)
   outmaps = Dict{Pair{edgetype(sites),QN},Dict{Vector{Op},Int}}()
 
-  #g = underlying_graph(sites)
-
   # Bond coefficients for incoming edge channels.
   # These become the "M" coefficient matrices that get SVD'd.
-  inbond_coefs = Dict(
-    e => Dict{QN,Vector{MatElem{coefficient_type}}}() for e in ordered_edges
-  )
+  #edge_matrix_type = SparseArrayDOK{coefficient_type}
+  edge_matrix_type = Vector{MatElem{coefficient_type}}
+  edge_data_eltype = Dict{QN,edge_matrix_type}
+  bond_coefs = DataGraph(sites; edge_data_eltype)
+  # Default initialize the coefs DataGraph
+  for e in edges(bond_coefs)
+    bond_coefs[e] = edge_data_eltype()
+  end
 
   # List of terms for which the coefficient has been added to a site factor
   site_coef_done = Prod{Op}[]
 
   # Temporary symbolic representation of TTN Hamiltonian
-  symbolic_ttn = Dict(
-    v => QNArrElem{Scaled{coefficient_type,Prod{Op}},degree(sites, v)}[] for
-    v in ordered_verts
-  )
+  vertex_data_eltype = Vector{QNArrElem{Scaled{coefficient_type,Prod{Op}}}}
+  symbolic_ttn = DataGraph(sites; vertex_data_eltype)
 
   # Build compressed finite state machine representation (symbolic_ttn)
-  for v in ordered_verts
+  for v in vertices(sites)
+    symbolic_ttn[v] = vertex_data_eltype()
     v_degree = degree(sites, v)
+
     # For every vertex, find all edges that contain this vertex
     # (align_and_reorder_edges makes the output of indicident edges match the
     #  direction and ordering match that of ordered_edges)
-    edges = align_and_reorder_edges(incident_edges(sites, v), ordered_edges)
+    v_edges = align_and_reorder_edges(incident_edges(sites, v), ordered_edges)
 
     # Use the corresponding ordering as index order for tensor elements at this site
-    dim_in = findfirst(e -> dst(e) == v, edges)
-    edge_in = (isnothing(dim_in) ? nothing : edges[dim_in])
-    dims_out = findall(e -> src(e) == v, edges)
-    edges_out = edges[dims_out]
+    dim_in = findfirst(e -> dst(e) == v, v_edges)
+    edge_in = (isnothing(dim_in) ? nothing : v_edges[dim_in])
+    dims_out = findall(e -> src(e) == v, v_edges)
+    edges_out = v_edges[dims_out]
 
     # For every site v' except v, determine the incident edge to v
     # that lies in the edge_path(v',v)
     # TODO: better way to make which_incident_edge below?
+    # TODO: instead of split at vertex, should
+    #       be able to traverse subtrees formed by cutting incoming bond
     subgraphs = split_at_vertex(sites, v)
-    _boundary_edges = [
-      only(boundary_edges(underlying_graph(sites), subgraph)) for subgraph in subgraphs
-    ]
-    _boundary_edges = align_edges(_boundary_edges, edges)
+    _boundary_edges = [only(boundary_edges(sites, subgraph)) for subgraph in subgraphs]
+    _boundary_edges = align_edges(_boundary_edges, v_edges)
     which_incident_edge = Dict(
       Iterators.flatten([
         subgraphs[i] .=> ((_boundary_edges[i]),) for i in eachindex(subgraphs)
       ]),
     )
 
```

Comment:
Given the target to simplify the code structure, I suggest removing this and reverting the dependent implementation below to the state of #116. For an efficient implementation of `vertex_path` the original implementation using 'vertex/edge_path` is unlikely to cause a significant overhead.

Reply:
Thanks for the link to #116. I will consider that, and a good plan here could be:
* revert the parts using `which_incident_edge` to use the older code, which looks quite nice and readable
* keep this code as a reference for performance, i.e. that we have to get back to at least this performance level in the future to consider the refactor 'done'
* try to find ways to either speed up calls involving `edge_path`, say, by using a different graph algorithm either inside `edge_path` or replacing `edge_path` with another algorithm altogether

We may end up making some new graph functions in the process which might end up repeating the `which_incident_edge` logic in the end, but the upshot is that it would be behind a more "self documenting" interface and properly pulled out as a separate function etc.